[6.x] Add `view other authors entries` permission

aerni opened the pull request:
This PR aims to solve the longstanding community request to restrict users' view access to other authors' entries. 

## Summary
- Adds new `view other authors {collection} entries` permission
- Supports entries with multiple authors
- Dynamically changes the `author` field visibility depending on the user's permissions
- Adds an update script to add the new permission to existing roles

## In Detail

### Permission & Policy
This PR adds a new `view other authors {collection} entries` permission and makes the existing author-related permissions its children. The updated policy is pretty straightforward. The tricky part was getting the query and filtering for the entries index listing and fieldtype working correctly. Thanks to @ryanmitchell for lending a hand.

Users don't require any of the `... other authors {collection} entries` permissions to:
- ... view/edit their own entries
- ... view/edit entries with blueprints that have no author field
- ... view/edit entries with multiple authors when they are part of the author array

### Author Field Visibility
This PR also dynamically changes the `author` field visibility and listing column depending on the user's permissions:

- The `author` field is `hidden` and the column removed from the index listing if the user has no `view` permission.
- The `author` field is  `read_only` if the user has `view` permission
- The `author` field is `visible` if the user has `edit` permission

The field visibility is changed in an `EntryBlueprintFound` listener so that users can override it if needed.

### Update Script
The new `view other authors {collection} entries` is added to roles that have an existing `edit other authors {collection} entries` permission. You will have to update `$this->isUpdatingTo('5.59');` to the correct version once you're ready to release this.

aerni committed:
Add permission

aerni committed:
Restrict other authors from viewing entries

aerni committed:
Only show author’s entries in listings

Also show entries with a blueprint that doesn’t have an author field

aerni committed:
Fix entries count

aerni committed:
Make Erin happy with shorter syntax

aerni committed:
Handle entries fieldtype

aerni committed:
Handle tree view

aerni committed:
🍺

aerni committed:
Move filtering to the controller

ryanmitchell commented on src/Fieldtypes/Entries.php:
```diff
@@ -144,6 +144,20 @@ public function getIndexItems($request)
+                    ->filter(fn ($blueprint) => ! $blueprint->hasField('author'))
+                    ->map->handle()->all();
+
+                    $query->where(fn ($query) => $query
```
As this can be across multiple collections and blueprint names are potentially the same across multiple collections (but different blueprints) you should `->where('collection', $collction->handle())` here too.

aerni committed:
Fix issue with multiple collections

aerni committed:
Dynamically change author field visibility

aerni committed:
Also remove author column

aerni commented:
@AtmoFX I think that's a valid feature request. But it might be a little overkill to add it in this PR. I'd like to keep it as simple as possible to make review easier. I see your request as a bigger overhaul of how other authors' entries are treated. But I'll leave it up to the core team to decide.

duncanmcclean commented:
It looks like some of the tests are failing. Are you able to take a look?

aerni commented:
Yes, will do. Didn't get to it yet 👍🏻

aerni committed:
Fix tests

aerni committed:
Add update script

Jubeki commented:
Good idea this PR, something I would like to use myself.

What happens if you configure the blueprint so that you can have multiple authors on the same entry? Would this still work as intended?

I would assume that all listed authors can view the entry, but not listed authors can not.

aerni committed:
Add support for multiple authors

aerni committed:
Abstract

aerni committed:
Fix tests

aerni committed:
Add tests

aerni committed:
Allow users to override field visibility

aerni committed:
Register event listener

aerni committed:
Make sure permissions apply to search

This also fixes an issue where the search would include unauthorized sites if the collection had a search index.

aerni committed:
Merge remote-tracking branch 'statamic/5.x' into feature/view-other-authors-entries

aerni committed:
Compare against string

aerni committed:
Use new whereHas

aerni committed:
Fix issue if there are no blueprints with author

aerni committed:
Fix test

aerni committed:
No need to add queries if there is no author field

aerni committed:
Make it more readable
